src/backends/interflop-mca-int/interflop_mca_int.c

```diff
@@ -0,0 +1,681 @@
+} mca_operations;
+
+static float _mca_binary32_binary_op(float a, float b, const mca_operations op,
+                                     void *context);
+static double _mca_binary64_binary_op(double a, double b,
+                                      const mca_operations op, void *context);
+
+/******************** MCA CONTROL FUNCTIONS *******************
+ * The following functions are used to set virtual precision and
+ * MCA mode of operation.
+ ***************************************************************/
+
+/* Set the mca mode */
+static void _set_mca_mode(const mcamode mode) {
+  if (mode >= _mcamode_end_) {
+    logger_error("--%s invalid value provided, must be one of: "
+                 "{ieee, mca, pb, rr}.",
+                 key_mode_str);
+  }
+  MCALIB_MODE = mode;
+}
+
+/* Set the virtual precision for binary32 */
+static void _set_mca_precision_binary32(const int precision) {
+  _set_precision(MCA, precision, &MCALIB_BINARY32_T, (float)0);
+}
+
+/* Set the virtual precision for binary64 */
+static void _set_mca_precision_binary64(const int precision) {
+  _set_precision(MCA, precision, &MCALIB_BINARY64_T, (double)0);
+}
+
+/******************** MCA RANDOM FUNCTIONS ********************
+ * The following functions are used to calculate the random
+ * perturbations used for MCA
+ ***************************************************************/
+
+/* global thread id access lock */
+static pthread_mutex_t global_tid_lock = PTHREAD_MUTEX_INITIALIZER;
+/* global thread identifier */
+static unsigned long long int global_tid = 0;
+
+/* helper data structure to centralize the data used for random number
+ * generation */
+static __thread rng_state_t rng_state;
+
+/* noise = rand * 2^(exp) */
+/* We can skip special cases since we never met them */
+/* Since we have exponent of float values, the result */
+/* is comprised between: */
+/* 127+127 = 254 < DOUBLE_EXP_MAX (1023)  */
+/* -126-24+-126-24 = -300 > DOUBLE_EXP_MIN (-1022) */
+static inline void _noise_binary64(double *x, const int exp,
+                                   rng_state_t *rng_state) {
+  uint64_t noise_mask, noise, noise_msb, noise_sign_mask;
+  uint32_t mask_shift_amount;
+  binary64 x_b64 = {.f64 = *x};
+
+  // amount by which to shift, when creating the mask for the noise
+  mask_shift_amount = 1 + DOUBLE_EXP_SIZE + exp;
+  // create the mask for the noise
+  noise_mask = ((uint64_t)DOUBLE_MASK_ONE) >> mask_shift_amount;
+
+  // generate a new random 64-bit integer
+  noise = _get_rand_uint64(rng_state, &global_tid_lock, &global_tid);
+
+  // extract the MSB of the noise
+  //  first extract the MSB
+  noise_msb = noise & (((uint64_t)1) << (DOUBLE_EXP_SIZE + DOUBLE_PMAN_SIZE));
+  //  next, shift the MSB all the way to the LSB position
+  noise_msb = noise_msb >> (DOUBLE_EXP_SIZE + DOUBLE_PMAN_SIZE);
+  // noise sign mask is used to create the one's complement of the masked noise
+  noise_sign_mask = (uint64_t)DOUBLE_MASK_ONE + (noise_msb ^ 1);
```

Comment:
I do not think we need to treat separately the sign bit.
Indeed the first bit of `noise` can act as a sign bit naturally.
When shifting the noise with >> (DOUBLE_EXP_SIZE + DOUBLE_PMAN_SIZE) the sign bit will be correctly extended.

It seems a quite simpler implementation is possible that only does:
```c
noise = _get_rand_uint64(rng_state, &global_tid_lock, &global_tid);
noise= noise >> (DOUBLE_EXP_SIZE + DOUBLE_PMAN_SIZE);
x_b64.u64 = x_b64.u64 + noise;
```
I am testing this on a separate test harness to check that it matches SR expected distributions on some simple inexact values and it works well.